### PR TITLE
Introduce lockedSource type to avoid concurrent use of rand.Source

### PR DESCRIPTION
**Pull request by RomanZavodskikh**
+ Introduced lockedSource type to avoid concurrent use of rand.Source
+ Added test calling lockedSource concurrently
+ Added benchmark for loadbalancer
+ Removed some unused fields from {random,roundRobin,consistentHash} structs
+ Preallocate slices
+ Do not use weightedSums slice

Updates https://github.com/zalando/skipper/issues/2346

### CODE REVIEW

File: loadbalancer/locked_rand.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package loadbalancer
+
+import (
+	"math/rand"
+	"sync"
+)
+
+type lockedRand struct {
```
[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply loadbalancer/locked_rand.go:8]: We may also add a common constructor to it like `rand.NewSource(time.Now().UnixNano())`

File: loadbalancer/algorithm.go [hunk 2]
```diff
@@ -145,13 +145,13 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, notFadingIndexes []int,
 type roundRobin struct {
 	mx               sync.Mutex
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:146]: Is mx still needed?

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:146]: Ah, ok. I guess we could also use evergrowing https://pkg.go.dev/sync/atomic#Int64 and get rid of the mutex similar to what we do here https://github.com/zalando/skipper/blob/e88336207452706b8fe92e3086a6b93b327dfbd5/ratelimit/ratelimit.go#L271-L275

**Comment by szuecs** [comment]
I think we should also add tests to cover this code part.

File: loadbalancer/algorithm.go [hunk 3]
```diff
@@ -143,17 +144,16 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, notFadingIndexes []int,
 }
 
 type roundRobin struct {
-	mx               sync.Mutex
-	index            int
+	index            int64
 	rnd              *rand.Rand
 	notFadingIndexes []int
 	fadingWeights    []float64
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:150]: Hm, looks like we have to guard notFadingIndexes and fadingWeights

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:150]: Yes, it makes sense to have mutex for every loadbalancer struct using `withFadeIn` function.
However, I'd like still to have `lockedSource` object not to forget to lock `rand.Source` in the future.

**Comment by RomanZavodskikh** [comment]
> I think we should also add tests to cover this code part.
@szuecs I've added tests for `locked_source.go` part. And it seems to me there's already `algorithm_test.go` present in the project. Do we need to add something more?

File: loadbalancer/algorithm.go (hunk 3 above)
Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:150]: @AlexanderYastrebov it looks like those fields are needed only for preallocation purposes, not to do actual writes to those fields.

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:150]: @AlexanderYastrebov withFadeIn writes to *copy* of it, like here:
```
rzavodskikh@ZALANDO-39125:/tmp$ cat copySlice.go 
package main

import (
	"fmt"
)

func foo(s []int) {
	s = append(s, 4)
}

func main() {
	a := []int{1, 2, 3}
	foo(a)

	fmt.Printf("a = %+v\n", a)
}

rzavodskikh@ZALANDO-39125:/tmp$ go build copySlice.go 
rzavodskikh@ZALANDO-39125:/tmp$ ./copySlice 
a = [1 2 3]
```

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:150]: https://go.dev/ref/spec#Appending_and_copying_slices

> If the capacity of s is not large enough to fit the additional values, append [allocates](https://go.dev/ref/spec#Allocation) a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array.

Since field capacity is pre-allocated the code writes to the same array even though it uses append (which might be a bit confusing, I agree).

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:150]: Yeah, I do understand that append inside `withFadeIn` function may write (and probably writes) to the same **underlying array**, that's true.
However, since caller passes **slice** to the `withFadeIn` function, after the return this **slice** will remain the same (even though there were writes to the **underlying array**, you can search by "The output might be confusing!" here: https://trstringer.com/golang-slice-references/ )

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:150]: The only clue that I see is comment `// preallocating frequently used slice`, so it seems to me those slices were intended to improve performance of `Apply`.
However, when I tried to benchmark this operation, I see no significant difference with such fields and without them:
```
Every 10.0s: benchstat old.txt new.txt                                                                                                                                      ZALANDO-39125: Wed Jun 21 16:30:37 2023

goos: linux
goarch: amd64
pkg: github.com/zalando/skipper/loadbalancer
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                                           │   old.txt   │              new.txt               │
                                           │   sec/op    │   sec/op     vs base               │
FadeIn/random,_10,_1_clients-16              830.2n ± 3%   871.6n ± 3%  +4.99% (p=0.000 n=74)
FadeIn/random,_10,_4_clients-16              1.147µ ± 8%   1.081µ ± 6%       ~ (p=0.084 n=74)
FadeIn/random,_10,_16_clients-16             1.261µ ± 4%   1.203µ ± 4%       ~ (p=0.169 n=74)
FadeIn/random,_10,_64_clients-16             1.044µ ± 4%   1.010µ ± 4%       ~ (p=0.084 n=74)
FadeIn/random,_10,_256_clients-16            1.071µ ± 3%   1.032µ ± 3%       ~ (p=0.139 n=74)
FadeIn/round-robin,_10,_1_clients-16         928.0n ± 3%   983.6n ± 3%  +6.00% (p=0.000 n=74)
FadeIn/round-robin,_10,_4_clients-16         762.3n ± 5%   778.2n ± 6%       ~ (p=0.527 n=74)
FadeIn/round-robin,_10,_16_clients-16        813.8n ± 6%   787.0n ± 4%       ~ (p=0.371 n=74)
FadeIn/round-robin,_10,_64_clients-16        633.2n ± 3%   618.3n ± 1%  -2.36% (p=0.043 n=74)
FadeIn/round-robin,_10,_256_clients-16       669.3n ± 5%   653.5n ± 2%       ~ (p=0.234 n=74)
FadeIn/consistent-hash,_10,_1_clients-16     1.169µ ± 3%   1.258µ ± 2%  +7.61% (p=0.000 n=74)
FadeIn/consistent-hash,_10,_4_clients-16     611.6n ± 2%   649.4n ± 4%  +6.17% (p=0.006 n=74)
FadeIn/consistent-hash,_10,_16_clients-16    525.1n ± 5%   545.9n ± 3%       ~ (p=0.234 n=74)
FadeIn/consistent-hash,_10,_64_clients-16    430.3n ± 2%   436.9n ± 3%       ~ (p=0.456 n=74)
FadeIn/consistent-hash,_10,_256_clients-16   448.9n ± 2%   450.3n ± 1%       ~ (p=0.533 n=74)
geomean                                      779.3n        782.4n       +0.40%
```

File: loadbalancer/locked_source_test.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	}
+}
+
+func TestLockedSourceForConcurrentUse(t *testing.T) {
+	s := newLockedSource()
+
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go loadTestLockedSource(s, 100000, &wg)
```
Comment (szuecs) [inline loadbalancer/locked_source_test.go:22]: normally you would do something like this to not pass &wg:
```
go func() {
    loadTestLockedSource(s, 100000)
    wg.Done()
}()
```

**Comment by szuecs** [comment]
@RomanZavodskikh Can you post pictures of our scripting to show that this looks also fine?

**Comment by RomanZavodskikh** [comment]
TL;DR All graph are essentially the same as in the master branch.

TestFadeIn_consistent-hash,_1
TestFadeIn_consistent-hash,_3
TestFadeIn_consistent-hash,_4
TestFadeIn_consistent-hash,_5
TestFadeIn_consistent-hash,_6
TestFadeIn_consistent-hash,_7
TestFadeIn_consistent-hash,_9
TestFadeIn_random,_1
TestFadeIn_random,_3
TestFadeIn_random,_4
TestFadeIn_random,_5
TestFadeIn_random,_6
TestFadeIn_random,_7
TestFadeIn_random,_9
TestFadeIn_round-robin,_1
TestFadeIn_round-robin,_3
TestFadeIn_round-robin,_4
TestFadeIn_round-robin,_5
TestFadeIn_round-robin,_6
TestFadeIn_round-robin,_7
TestFadeIn_round-robin,_9

**Comment by szuecs** [comment]
:+1:

File: loadbalancer/algorithm.go [hunk 5]
```diff
@@ -119,6 +120,7 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, notFadingIndexes []int,
 	if rnd.Float64() < f {
 		return ep[choice]
 	}
+	notFadingIndexes := []int{}
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:123]: Here we can pre-allocate len(ep) capacity

File: loadbalancer/algorithm.go [hunk 6]
```diff
@@ -99,14 +100,14 @@ func shiftToRemaining(rnd *rand.Rand, ctx *routing.LBContext, wi []int, wf []flo
 	// the perspective of the oldest endpoint, this is temporarily the same as if
 	// there was no fade-in.
 	if len(notFadingIndexes) == 0 {
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:102]: This is always true, isn't it?

File: loadbalancer/algorithm.go [hunk 7]
```diff
@@ -67,9 +68,9 @@ func fadeIn(now time.Time, duration time.Duration, exponent float64, detected ti
 	return math.Pow(float64(rel)/float64(duration), exponent)
 }
 
-func shiftWeighted(rnd *rand.Rand, ctx *routing.LBContext, w []float64, now time.Time) routing.LBEndpoint {
+func shiftWeighted(rnd *rand.Rand, ctx *routing.LBContext, now time.Time) routing.LBEndpoint {
 	var sum float64
-	weightSums := w
+	weightSums := []float64{}
```
[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:73]: Also this now introduces an allocation for each request

[the remark this answers is not in the document]
Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:73]: It sound like a really good idea, but by some reason performance drops dramatically. So I'd like **not** to allocate the capacity right away.

```
rzavodskikh@ZALANDO-39125:~/skipper$ git diff origin/lockedRand lockedRand
diff --git a/loadbalancer/algorithm.go b/loadbalancer/algorithm.go
index 919b45bf..6c9d55f9 100644
--- a/loadbalancer/algorithm.go
+++ b/loadbalancer/algorithm.go
@@ -70,7 +70,7 @@ func fadeIn(now time.Time, duration time.Duration, exponent float64, detected ti
 
 func shiftWeighted(rnd *rand.Rand, ctx *routing.LBContext, now time.Time) routing.LBEndpoint {
        var sum float64
-       weightSums := []float64{}
+       weightSums := make([]float64, 0, len(ctx.Route.LBEndpoints))
        rt := ctx.Route
        ep := ctx.Route.LBEndpoints
        for _, epi := range ep {
@@ -120,7 +120,7 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, choice int, algo routing
        if rnd.Float64() < f {
                return ep[choice]
        }
-       notFadingIndexes := []int{}
+       notFadingIndexes := make([]int, 0, len(ep))
        for i := 0; i < len(ep); i++ {
                if _, fadingIn := fadeInState(now, ctx.Route.LBFadeInDuration, ep[i].Detected); !fadingIn {
                        notFadingIndexes = append(notFadingIndexes, i)

```

```
goos: linux
goarch: amd64
pkg: github.com/zalando/skipper/loadbalancer
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                                           │   old.txt    │               new.txt                │
                                           │    sec/op    │    sec/op     vs base                │
FadeIn/random,_10,_1_clients-16              854.8n ±  8%   1506.0n ± 3%  +76.18% (p=0.000 n=17)
FadeIn/random,_10,_4_clients-16              748.0n ±  7%    981.1n ± 6%  +31.16% (p=0.000 n=17)
FadeIn/random,_10,_16_clients-16             881.6n ±  6%   1078.0n ± 3%  +22.28% (p=0.000 n=17)
FadeIn/random,_10,_64_clients-16             1.044µ ±  2%    1.213µ ± 2%  +16.19% (p=0.000 n=17)
FadeIn/random,_10,_256_clients-16            1.074µ ±  4%    1.212µ ± 3%  +12.85% (p=0.000 n=17)
FadeIn/round-robin,_10,_1_clients-16         863.6n ±  4%   1513.0n ± 2%  +75.20% (p=0.000 n=17)
FadeIn/round-robin,_10,_4_clients-16         630.0n ± 16%    851.9n ± 5%  +35.22% (p=0.000 n=17)
FadeIn/round-robin,_10,_16_clients-16        626.2n ±  6%    782.6n ± 5%  +24.98% (p=0.000 n=17)
FadeIn/round-robin,_10,_64_clients-16        662.3n ±  4%    858.5n ± 2%  +29.62% (p=0.000 n=17)
FadeIn/round-robin,_10,_256_clients-16       677.2n ±  2%    854.6n ± 2%  +26.20% (p=0.000 n=17)
FadeIn/consistent-hash,_10,_1_clients-16     1.085µ ±  6%    1.959µ ± 3%  +80.55% (p=0.000 n=17)
FadeIn/consistent-hash,_10,_4_clients-16     576.1n ±  6%    839.1n ± 3%  +45.65% (p=0.000 n=17)
FadeIn/consistent-hash,_10,_16_clients-16    471.5n ± 10%    618.5n ± 3%  +31.18% (p=0.000 n=17)
FadeIn/consistent-hash,_10,_64_clients-16    493.2n ±  1%    655.2n ± 3%  +32.85% (p=0.000 n=17)
FadeIn/consistent-hash,_10,_256_clients-16   510.6n ±  2%    646.7n ± 3%  +26.65% (p=0.000 n=17)
geomean                                      719.5n          980.7n       +36.30%
```

File: loadbalancer/algorithm.go (hunk 5 above)
Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:123]: https://github.com/zalando/skipper/pull/2412#discussion_r1238267164

File: loadbalancer/algorithm.go (hunk 6 above)
Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:102]: Yes.

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:102]: It's not.

File: loadbalancer/algorithm.go [hunk 8]
```diff
@@ -168,32 +163,23 @@ func (r *roundRobin) Apply(ctx *routing.LBContext) routing.LBEndpoint {
...
-	defer r.mx.Unlock()
-	r.index = (r.index + 1) % len(ctx.Route.LBEndpoints)
+	index := int(atomic.AddInt64(&r.index, 1) % int64(len(ctx.Route.LBEndpoints)))
 
 	if ctx.Route.LBFadeInDuration <= 0 {
-		return ctx.Route.LBEndpoints[r.index]
+		return ctx.Route.LBEndpoints[index]
 	}
 
-	return withFadeIn(r.rnd, ctx, r.notFadingIndexes, r.index, r)
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:179]: I think we misunderstood the idea of this `r.notFadingIndexes`: yes here we pass a slice by value and its length is always zero but capacity is equal to `len(endpoints)` (see constructor). 
`withFadeIn` writes to its underlying array and then passes to `shiftToRemaining` which uses values from underlying array.
So the idea of `r.notFadingIndexes` is to act like a scratch buffer to avoid allocations on each request (https://github.com/zalando/skipper/pull/2412/files#r1238234491).
The real problem is that access to this scratch buffer is not synchronized.

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:179]: The same is true for `a.fadingWeights` - it is also used as a scratch buffer.

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:179]:
```
access to this scratch buffer is not synchronized.
```
I still don't understand why should we synchronize access to the struct field we never write to.

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:179]: Another thought is that it's possible to synchronize access by per-loadbalancer lock like this, but this way performance will suffer while concurrent access, much more that with suboptimal memory (re)allocation.

```
+       ch.mx.Lock()
+       defer ch.mx.Unlock()
        return withFadeIn(ch.rand, ctx, ch.notFadingIndexes, choice, ch)
 }
```

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:179]:
```
 is to act like a scratch buffer to avoid allocations on each request
```
One more thought - this idea sounds legit in theory, but I haven't seen any performance advance by using such scratch buffers.

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:179]: > One more thought - this idea sounds legit in theory, but in practice I haven't seen any performance advance by using such scratch buffers.

I think you saw https://github.com/zalando/skipper/pull/2412/files#r1238267164

File: loadbalancer/algorithm.go (hunk 5 above)
Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:123]: I think we should not preallocate here.

Actually, `notFadingIndexes` length depends on the number of non-fading endpoints which may change over time. So, capacity to allocate with is not known.
However, I tried to preallocate with `make([]int, 0, len(ep))` here and tested the performance, I see it's degrading here:

```
                                           │ lockedRand.txt │           preallocate2.txt           │
                                           │     sec/op     │    sec/op     vs base                │
FadeIn/random,_10,_1_clients-16                894.0n ±  5%   1619.0n ± 5%  +81.10% (p=0.000 n=26)
FadeIn/random,_11,_1_clients-16                403.8n ±  9%    417.8n ± 8%        ~ (p=0.304 n=26)
FadeIn/random,_10,_4_clients-16                879.0n ± 12%   1031.5n ± 6%  +17.34% (p=0.000 n=26)
FadeIn/random,_11,_4_clients-16                548.4n ±  7%    592.6n ± 6%        ~ (p=0.102 n=26)
FadeIn/random,_10,_16_clients-16               914.6n ±  3%   1101.0n ± 2%  +20.38% (p=0.000 n=26)
FadeIn/random,_11,_16_clients-16               624.3n ±  6%    621.0n ± 5%        ~ (p=0.710 n=26)
FadeIn/random,_10,_64_clients-16               1.003µ ±  3%    1.183µ ± 1%  +18.00% (p=0.000 n=26)
FadeIn/random,_11,_64_clients-16               700.5n ±  4%    704.1n ± 6%        ~ (p=0.498 n=26)
FadeIn/random,_10,_256_clients-16              1.016µ ±  4%    1.156µ ± 1%  +13.84% (p=0.000 n=26)
FadeIn/random,_11,_256_clients-16              703.8n ±  2%    704.0n ± 5%        ~ (p=0.770 n=26)
FadeIn/round-robin,_10,_1_clients-16           935.7n ±  5%   1704.5n ± 4%  +82.16% (p=0.000 n=26)
FadeIn/round-robin,_11,_1_clients-16           383.0n ±  6%    432.8n ± 5%  +13.00% (p=0.000 n=26)
FadeIn/round-robin,_10,_4_clients-16           629.8n ± 10%    905.1n ± 5%  +43.72% (p=0.000 n=26)
FadeIn/round-robin,_11,_4_clients-16           341.4n ± 17%    379.7n ± 8%  +11.23% (p=0.047 n=26)
FadeIn/round-robin,_10,_16_clients-16          624.9n ±  7%    775.5n ± 4%  +24.10% (p=0.000 n=26)
FadeIn/round-robin,_11,_16_clients-16          350.7n ±  7%    377.6n ± 8%        ~ (p=0.069 n=26)
FadeIn/round-robin,_10,_64_clients-16          650.5n ±  3%    819.0n ± 3%  +25.91% (p=0.000 n=26)
FadeIn/round-robin,_11,_64_clients-16          383.2n ±  1%    384.4n ± 6%        ~ (p=0.840 n=26)
FadeIn/round-robin,_10,_256_clients-16         651.6n ±  1%    815.8n ± 2%  +25.21% (p=0.000 n=26)
FadeIn/round-robin,_11,_256_clients-16         386.9n ±  1%    390.6n ± 2%        ~ (p=0.161 n=26)
FadeIn/consistent-hash,_10,_1_clients-16       1.185µ ±  5%    2.211µ ± 3%  +86.54% (p=0.000 n=26)
FadeIn/consistent-hash,_11,_1_clients-16       627.8n ±  9%    663.2n ± 6%        ~ (p=0.234 n=26)
FadeIn/consistent-hash,_10,_4_clients-16       568.9n ±  4%    873.6n ± 4%  +53.55% (p=0.000 n=26)
FadeIn/consistent-hash,_11,_4_clients-16       452.2n ±  8%    433.4n ± 8%        ~ (p=0.426 n=26)
FadeIn/consistent-hash,_10,_16_clients-16      462.2n ±  9%    598.2n ± 2%  +29.41% (p=0.000 n=26)
FadeIn/consistent-hash,_11,_16_clients-16      430.4n ±  5%    429.6n ± 5%        ~ (p=0.602 n=26)
FadeIn/consistent-hash,_10,_64_clients-16      483.9n ±  1%    636.8n ± 4%  +31.60% (p=0.000 n=26)
FadeIn/consistent-hash,_11,_64_clients-16      430.9n ±  1%    428.9n ± 5%        ~ (p=0.558 n=26)
FadeIn/consistent-hash,_10,_256_clients-16     497.3n ±  2%    629.1n ± 2%  +26.50% (p=0.000 n=26)
FadeIn/consistent-hash,_11,_256_clients-16     433.8n ±  5%    433.4n ± 7%        ~ (p=0.680 n=26)
geomean                                        582.4n          690.8n       +18.61%
```

File: loadbalancer/algorithm.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go]: As for `weightSums`, we may or we may not preallocate it. It's only `notFadingIndexes` which affects performance.

```
goos: linux
goarch: amd64
pkg: github.com/zalando/skipper/loadbalancer
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                                           │ lockedRand.txt │           preallocate.txt           │
                                           │     sec/op     │    sec/op     vs base               │
FadeIn/random,_10,_1_clients-16                894.0n ±  5%   867.0n ±  6%       ~ (p=0.279 n=26)
FadeIn/random,_11,_1_clients-16                403.8n ±  9%   389.2n ±  9%       ~ (p=0.721 n=26)
FadeIn/random,_10,_4_clients-16                879.0n ± 12%   791.4n ± 12%       ~ (p=0.080 n=26)
FadeIn/random,_11,_4_clients-16                548.4n ±  7%   556.0n ±  5%       ~ (p=0.978 n=26)
FadeIn/random,_10,_16_clients-16               914.6n ±  3%   959.9n ±  4%  +4.95% (p=0.016 n=26)
FadeIn/random,_11,_16_clients-16               624.3n ±  6%   645.2n ±  7%       ~ (p=0.295 n=26)
FadeIn/random,_10,_64_clients-16               1.003µ ±  3%   1.005µ ±  3%       ~ (p=0.770 n=26)
FadeIn/random,_11,_64_clients-16               700.5n ±  4%   700.9n ±  4%       ~ (p=0.589 n=26)
FadeIn/random,_10,_256_clients-16              1.016µ ±  4%   1.019µ ±  2%       ~ (p=0.634 n=26)
FadeIn/random,_11,_256_clients-16              703.8n ±  2%   705.6n ±  1%       ~ (p=0.194 n=26)
FadeIn/round-robin,_10,_1_clients-16           935.7n ±  5%   904.8n ±  7%       ~ (p=0.388 n=26)
FadeIn/round-robin,_11,_1_clients-16           383.0n ±  6%   400.3n ±  7%       ~ (p=0.304 n=26)
FadeIn/round-robin,_10,_4_clients-16           629.8n ± 10%   646.0n ±  8%       ~ (p=0.826 n=26)
FadeIn/round-robin,_11,_4_clients-16           341.4n ± 17%   376.2n ±  9%       ~ (p=0.146 n=26)
FadeIn/round-robin,_10,_16_clients-16          624.9n ±  7%   628.2n ±  4%       ~ (p=0.567 n=26)
FadeIn/round-robin,_11,_16_clients-16          350.7n ±  7%   376.6n ±  7%       ~ (p=0.070 n=26)
FadeIn/round-robin,_10,_64_clients-16          650.5n ±  3%   644.9n ±  3%       ~ (p=0.340 n=26)
FadeIn/round-robin,_11,_64_clients-16          383.2n ±  1%   383.2n ±  7%       ~ (p=0.876 n=26)
FadeIn/round-robin,_10,_256_clients-16         651.6n ±  1%   650.9n ±  1%       ~ (p=0.993 n=26)
FadeIn/round-robin,_11,_256_clients-16         386.9n ±  1%   387.1n ±  1%       ~ (p=0.763 n=26)
FadeIn/consistent-hash,_10,_1_clients-16       1.185µ ±  5%   1.167µ ±  4%       ~ (p=0.674 n=26)
FadeIn/consistent-hash,_11,_1_clients-16       627.8n ±  9%   604.6n ±  5%       ~ (p=0.605 n=26)
FadeIn/consistent-hash,_10,_4_clients-16       568.9n ±  4%   568.5n ±  5%       ~ (p=0.618 n=26)
FadeIn/consistent-hash,_11,_4_clients-16       452.2n ±  8%   449.6n ± 11%       ~ (p=0.805 n=26)
FadeIn/consistent-hash,_10,_16_clients-16      462.2n ±  9%   464.4n ±  3%       ~ (p=0.577 n=26)
FadeIn/consistent-hash,_11,_16_clients-16      430.4n ±  5%   430.6n ±  4%       ~ (p=0.504 n=26)
FadeIn/consistent-hash,_10,_64_clients-16      483.9n ±  1%   481.0n ±  2%       ~ (p=0.426 n=26)
FadeIn/consistent-hash,_11,_64_clients-16      430.9n ±  1%   430.5n ±  1%       ~ (p=0.641 n=26)
FadeIn/consistent-hash,_10,_256_clients-16     497.3n ±  2%   493.6n ±  1%       ~ (p=0.464 n=26)
FadeIn/consistent-hash,_11,_256_clients-16     433.8n ±  5%   433.2n ±  1%       ~ (p=0.370 n=26)
geomean                                        582.4n         583.6n        +0.21%
```

**Comment by AlexanderYastrebov** [comment]
:+1:

File: loadbalancer/algorithm.go [hunk 9]
```diff
@@ -127,9 +129,9 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, notFadingIndexes []int,
 
 	switch a := algo.(type) {
 	case *roundRobin:
-		return shiftToRemaining(a.rnd, ctx, notFadingIndexes, a.fadingWeights, now)
+		return shiftToRemaining(a.rnd, ctx, notFadingIndexes, now)
```
Comment (MustafaSaber) [inline loadbalancer/algorithm.go:132]: Maybe make `*rand.Rand` field name consistent in all algorithms?

Comment (MustafaSaber) [reply loadbalancer/algorithm.go:132]: but that's just a nitpick

Comment (RomanZavodskikh) [reply loadbalancer/algorithm.go:132]: Resolved. :-)

**Comment by MustafaSaber** [comment]
👍

**Comment by RomanZavodskikh** [comment]
:+1:
